src/library_pthread.js

```diff
@@ -319,18 +319,18 @@ var LibraryPThread = {
               err('Thread ' + d.threadId + ': ' + d.text);
             } else if (d.cmd === 'alert') {
               alert('Thread ' + d.threadId + ': ' + d.text);
-            } else if (d.cmd === 'exit') {
-              // Thread is exiting, no-op here
             } else if (d.cmd === 'exitProcess') {
               // A pthread has requested to exit the whole application process (runtime).
               Module['noExitRuntime'] = false;
               exit(d.returnCode);
-            } else if (d.cmd === 'cancelDone') {
-              PThread.freeThreadData(worker.pthread);
-              worker.pthread = undefined; // Detach the worker from the pthread object, and return it to the worker pool as an unused worker.
-              PThread.unusedWorkerPool.push(worker);
-              // TODO: Free if detached.
-              PThread.runningWorkers.splice(PThread.runningWorkers.indexOf(worker.pthread), 1); // Not a running Worker anymore.
+            } else if (d.cmd === 'exit' || d.cmd === 'cancelDone') {
+              var detached = Atomics.load(HEAPU32, (d.threadId + {{{ C_STRUCTS.pthread.detached }}} ) >> 2);
```

Comment:
Don't you want to maintain the old behaviour when `cancelDone` is the command?

What about making this a function and calling it unconditionally in `cancelDone` and conditionally during `exit`.

(I'm not an expert on this part of the code so I could well be wrong here, but we seems that you are change the behaviour on cancelDone.. also cancelDone doesn't have threadId property).

Reply:
Yeah making that a function might be a smarter idea.